flocker/node/agents/test/blockdevicefactory.py

```diff
@@ -0,0 +1,192 @@
+the current execution environment.
+
+This depends on a ``CLOUD_CONFIG_FILE`` environment variable being set.
+
+See `acceptance testing <acceptance-testing>`_ for details.
+
+.. code-block:: python
+
+    from .blockdevicefactory import ProviderType, get_blockdeviceapi
+
+    api = get_blockdeviceapi(ProviderType.openstack)
+    volume = api.create_volume(...)
+
+"""
+
+from os import environ
+from uuid import uuid4
+
+from yaml import safe_load
+
+from twisted.trial.unittest import SkipTest
+from twisted.python.constants import Names, NamedConstant
+
+from keystoneclient_rackspace.v2_0 import RackspaceAuth
+from keystoneclient.session import Session
+
+from cinderclient.client import Client as CinderClient
+from novaclient.client import Client as NovaClient
+
+from ..cinder import CinderBlockDeviceAPI
+from ..ebs import EBSBlockDeviceAPI, ec2_client
+from ..test.test_blockdevice import detach_destroy_volumes
+
+
+# The Rackspace authentication endpoint
+# See http://docs.rackspace.com/cbs/api/v1.0/cbs-devguide/content/Authentication-d1e647.html # noqa
+RACKSPACE_AUTH_URL = "https://identity.api.rackspacecloud.com/v2.0"
+
+
+class ConfigMissing(Exception):
+    """
+    The cloud configuration could not be found.
+    """
+
+
+class ProviderType(Names):
+    """
+    Kinds of compute/storage cloud providers for which this module is able to
+    build ``IBlockDeviceAPI`` providers.
+    """
+    openstack = NamedConstant()
+    aws = NamedConstant()
+
+
+def get_blockdeviceapi(provider):
+    """
+    Validate and load cloud provider's yml config file.
+    Default to ``~/acceptance.yml`` in the current user home directory, since
+    that's where buildbot puts its acceptance test credentials file.
+    """
+    cls, args = get_blockdeviceapi_args(provider)
+    return cls(**args)
+
+
+def get_blockdeviceapi_args(provider):
+    """
+    Get initializer arguments suitable for use in the instantiation of an
+    ``IBlockDeviceAPI`` implementation compatible with the given provider.
+
+    :param provider: A provider type the ``IBlockDeviceAPI`` is to be
+        compatible with.  A value from ``ProviderType``.
+
+    :raises: ``ConfigMissing`` if a ``CLOUD_CONFIG_FILE`` was not set and the
+        default config file could not be read.
+
+    :return: A ``dict`` that can initialize the matching implementation.
+    """
+    config_file_path = environ.get('CLOUD_CONFIG_FILE')
+    if config_file_path is None:
+        raise ConfigMissing(
+            'Supply the path to a cloud credentials file '
+            'using the CLOUD_CONFIG_FILE environment variable. '
+            'See: '
+            'https://docs.clusterhq.com/en/latest/gettinginvolved/acceptance-testing.html '  # noqa
+            'for details of the expected format.'
+        )
+
+    with open(config_file_path) as config_file:
+        config = safe_load(config_file.read())
+
+    section = config[provider.name]
```

Comment:
One constraint on us right now, unfortunately, is that our PistonCloud testig environment is ... shared.  I think we need to avoid sending *all* of our API keys to that slave.  I think that means we need to send a subset of the acceptance.yml.  Maybe we could have a quick discussion about our options for this sometime today.